quickwit-search/src/root.rs

```diff
@@ -0,0 +1,353 @@
+        .unwrap_or(&Vec::new())
+        .to_vec();
+    // Make the leaf search results after distributed search.
+    let leaf_search_result = LeafSearchResult {
+        num_hits,
+        partial_hits,
+    };
+
+    // Create a hash map of PartialHit with split as a key.
+    let mut partial_hits_map: HashMap<String, Vec<PartialHit>> = HashMap::new();
+    for partial_hit in leaf_search_result.partial_hits.iter() {
+        partial_hits_map
+            .entry(partial_hit.split.clone())
+            .or_insert_with(Vec::new)
+            .push(partial_hit.clone());
+    }
+
+    // Create a job for fetching docs and assign the splits that the node is responsible for based on the job.
+    let fetch_docs_jobs: Vec<Job> = partial_hits_map
+        .keys()
+        .map(|split| {
+            // TODO: Change to a better way that does not use unwrap().
+            let split_metadata = split_metadata_map.get(split).unwrap();
+            Job {
+                split: split.clone(),
+                cost: compute_split_cost(split_metadata),
+            }
+        })
+        .collect();
+    let assigned_fetch_docs_jobs = client_pool.assign_jobs(fetch_docs_jobs).await?;
+
+    // Perform the fetch docs phese.
+    let mut fetch_docs_handles: Vec<JoinHandle<anyhow::Result<FetchDocsResult>>> = Vec::new();
+    for (addr, jobs) in assigned_fetch_docs_jobs {
+        for job in jobs {
+            let mut search_client = if let Some(client) = clients.get(&addr) {
+                client.clone()
+            } else {
+                return Err(anyhow::anyhow!(
+                    "There is no search client for {} in the search client pool.",
+                    addr
+                ));
+            };
+
+            // TODO: Change to a better way that does not use unwrap().
+            let partial_hits = partial_hits_map.get(&job.split).unwrap().clone();
+            let fetch_docs_request = FetchDocsRequest {
+                partial_hits,
+                index_id: search_request.index_id.clone(),
+            };
+            let handle = tokio::spawn(async move {
+                match search_client.fetch_docs(fetch_docs_request).await {
+                    Ok(resp) => Ok(resp.into_inner()),
+                    Err(err) => Err(anyhow::anyhow!("Failed to fetch docs due to {:?}", err)),
+                }
+            });
+            fetch_docs_handles.push(handle);
+        }
+    }
+    let fetch_docs_responses = futures::future::try_join_all(fetch_docs_handles).await?;
+
+    let mut hits: Vec<Hit> = Vec::new();
+    for response in fetch_docs_responses {
+        match response {
+            Ok(fetch_docs_result) => {
+                hits.extend(fetch_docs_result.hits);
+            }
+            Err(err) => error!(err=?err),
+        }
+    }
+
+    let elapsed = start_instant.elapsed();
+
+    Ok(SearchResult {
+        num_hits: leaf_search_result.num_hits,
+        hits: hits,
```

Comment:
I think the hits needs to be sorted (!?).

Can you run `cargo clippy`?

Reply:
Sorted.
Please see https://github.com/quickwit-inc/quickwit/pull/163/commits/73aec4373b13726b81f7702d904fafd8eddfb0f6